### PR TITLE
Support multiple reward tokens, support native token as a reward token

### PR DESCRIPTION
**Pull request by manhlx3006**

### CODE REVIEW

File: contracts/liquidityMining/KyberRewardLocker.sol [hunk 1]
```diff
@@ -250,7 +301,12 @@ contract KyberRewardLocker is IKyberRewardLocker, PermissionAdmin {
       totalVesting
     );
 
-    token.safeTransfer(msg.sender, totalVesting);
+    if (token == IERC20Ext(0)) {
```
Comment (ducquangkstn) [inline contracts/liquidityMining/KyberRewardLocker.sol:304]: Another way to do this is keep weth in RewardLocker and FairLaunch.
At '_completeVesting' function, check if this token is weth, if yes then unwarp it, and transfer to user. (So we only need minimum changes)

File: contracts/liquidityMining/KyberFairLaunch.sol [hunk 2]
```diff
@@ -113,69 +126,83 @@ contract KyberFairLaunch is IKyberFairLaunch, PermissionAdmin, ReentrancyGuard {
...
   }
 
   /**
    * @dev allow admin to withdraw only reward token
    */
-  function adminWithdraw(uint256 amount) external onlyAdmin {
-    rewardToken.safeTransfer(msg.sender, amount);
+  function adminWithdraw(uint256 rewardTokenIndex, uint256 amount) external onlyAdmin {
+    IERC20Ext rewardToken = IERC20Ext(rewardTokens[rewardTokenIndex]);
+    if (rewardToken == IERC20Ext(0)) {
```
Comment (ducquangkstn) [inline contracts/liquidityMining/KyberFairLaunch.sol:148]: Without a `receive` method, I don't think this contract can receive native token (eth)

**Review by ducquangkstn — COMMENTED** [review]
Some drawbacks of this approach:
- rewardTokens are immutable
- each pool must follow the same startBlock, endBlock

File: contracts/liquidityMining/KyberRewardLocker.sol (hunk 1 above)
Comment (manhlx3006) [reply contracts/liquidityMining/KyberRewardLocker.sol:304]: yeah thought about it as well, fairlaunch only handles erc20, but don't think it is too much changes to support native token here

**Comment by manhlx3006** [comment]
1. rewardTokens is reference type, thus, can not be marked as `immutable` (which only works for value types).
2. each pool can still have different start and end blocks, esp when we only renew a few pools

File: contracts/liquidityMining/KyberFairLaunch.sol [hunk 3]
```diff
@@ -327,20 +371,35 @@ contract KyberFairLaunch is IKyberFairLaunch, PermissionAdmin, ReentrancyGuard {
...
-    uint256 _accRewardPerShare = pool.accRewardPerShare;
     uint256 _totalStake = pool.totalStake;
-    uint32 lastAccountedBlock = _lastAccountedRewardBlock(_pid);
-    if (lastAccountedBlock > pool.lastRewardBlock && _totalStake != 0) {
-      uint256 reward = uint256(lastAccountedBlock - pool.lastRewardBlock).mul(pool.rewardPerBlock);
-      _accRewardPerShare = _accRewardPerShare.add(reward.mul(PRECISION) / _totalStake);
+    uint256 _poolLastRewardBlock = pool.lastRewardBlock;
+    for(uint256 i = 0; i < rTokens.length; i++) {
+      uint256 _accRewardPerShare = pool.poolRewardData[rTokens[i]].accRewardPerShare;
+      uint32 lastAccountedBlock = _lastAccountedRewardBlock(_pid);
```
Comment (ducquangkstn) [inline contracts/liquidityMining/KyberFairLaunch.sol:384]: Pls move this outside of loop block

File: contracts/liquidityMining/KyberRewardLocker.sol [hunk 4]
```diff
@@ -94,15 +94,31 @@ contract KyberRewardLocker is IKyberRewardLocker, PermissionAdmin {
...
     uint256 schedulesLength = schedules.length;
     uint256 endBlock = startBlock.add(vestingDurationPerToken[token]);
+
+    // combine with the last schedule if they have the same start & end blocks
+    if (schedulesLength > 0) {
+      VestingSchedule storage lastSchedule = schedules.data[schedulesLength - 1];
+      if (lastSchedule.startBlock == startBlock && lastSchedule.endBlock == endBlock) {
+        lastSchedule.quantity = uint256(lastSchedule.quantity).add(quantity).toUint128();
+        accountEscrowedBalance[account][token] = accountEscrowedBalance[account][token].add(quantity);
+        emit VestingEntryCreated(token, account, startBlock, endBlock, quantity, schedulesLength - 1);
```
Comment (ducquangkstn) [inline contracts/liquidityMining/KyberRewardLocker.sol:117]: This can be changed to `VestingEntryQueued`

**Review by ducquangkstn — APPROVED** [review]
Generally, the code looks ok to me

If having time, we may can change from `userRewardData` to `mapping (uint256 => UserRewardData) userRewardData`
The key is index of reward token.

File: contracts/liquidityMining/KyberFairLaunch.sol (hunk 3 above)
Comment (manhlx3006) [reply contracts/liquidityMining/KyberFairLaunch.sol:384]: updated

File: contracts/liquidityMining/KyberRewardLocker.sol (hunk 4 above)
Comment (manhlx3006) [reply contracts/liquidityMining/KyberRewardLocker.sol:117]: added new event, will need to state it clearly when reading data from events

**Review by ducquangkstn — APPROVED** [review]

**Pull request merged**